metronome/core/src/io/iohk/metronome/core/messages/RPCSupport.scala

```diff
@@ -0,0 +1,110 @@
+      (_, _) => ().pure[F]
+    protected val responseIgnored: (K, Response, Option[Throwable]) => F[Unit] =
+      (_, _, _) => ().pure[F]
+
+    /** Send a request to the peer and track the response.
+      *
+      * Returns `None` if we're not connected or the request times out.
+      */
+    protected def sendRequest[Req <: Request, Res <: Response](
+        to: K,
+        mkRequest: RPCMessageCompanion#RequestId => Req
+    )(implicit
+        ev: RPCPair.Aux[Req, Res],
+        ct: ClassTag[Res]
+    ): F[Option[Res]] =
+      for {
+        requestId <- requestId
+        request = mkRequest(requestId)
+        join     <- rpcTracker.register[Req, Res](request)
+        _        <- sendRequest(to, request)
+        maybeRes <- join
+        _        <- requestTimeout(to, request).whenA(maybeRes.isEmpty)
+      } yield maybeRes
+
+    /** Try to complete a request when the response arrives. */
+    protected def receiveResponse[Res <: Response](
+        from: K,
+        response: Res
+    ): F[Unit] =
+      rpcTracker.complete(response).flatMap {
+        case Right(ok) =>
+          responseIgnored(from, response, None).whenA(!ok)
+        case Left(ex) =>
+          responseIgnored(from, response, Some(ex))
+      }
+  }
+
+  /** Identical to `RPCSupport.Remote` in functionality but without support for remote address. */
+  abstract class Local[
+      F[_]: Sync,
+      M,
+      Request <: RPCMessageCompanion#Request,
+      Response <: RPCMessageCompanion#Response
+  ](
+      rpcTracker: RPCTracker[F, M],
+      requestId: F[RPCMessageCompanion#RequestId]
+  )(implicit ev1: Request <:< M, ev2: Response <:< M) {
+
+    protected def sendRequest: Request => F[Unit]
+
+    protected val requestTimeout: Request => F[Unit] =
+      _ => ().pure[F]
+    protected val responseIgnored: (Response, Option[Throwable]) => F[Unit] =
+      (_, _) => ().pure[F]
+
+    /** Send a request and track the response.
+      *
+      * Returns `None` if we're not connected or the request times out.
+      */
+    protected def sendRequest[Req <: Request, Res <: Response](
+        mkRequest: RPCMessageCompanion#RequestId => Req
+    )(implicit
+        ev: RPCPair.Aux[Req, Res],
+        ct: ClassTag[Res]
+    ): F[Option[Res]] =
+      for {
+        requestId <- requestId
+        request = mkRequest(requestId)
+        join     <- rpcTracker.register[Req, Res](request)
+        _        <- sendRequest(request)
+        maybeRes <- join
+        _        <- requestTimeout(request).whenA(maybeRes.isEmpty)
+      } yield maybeRes
+
+    /** Try to complete a request when the response arrives. */
+    protected def receiveResponse[Res <: Response](
+        response: Res
+    ): F[Unit] =
+      rpcTracker.complete(response).flatMap {
+        case Right(ok) =>
+          responseIgnored(response, None).whenA(!ok)
+        case Left(ex) =>
+          responseIgnored(response, Some(ex))
+      }
+  }
```

Comment:
Remote and Local seems to be very similar to each other. Yet I do not see ATM how one could by more DRY. Possibly Local is just a special case of Remote when `to` argument is self.

Reply:
My first attempt was where `Local` just inherited `Remote` using `Unit` for `PKey` and passing `()`, but the problem was that with inheritance the classes that used them got both versions of the `sendRequest` methods, which didn't feel right. 

I thought about not using inheritance but some component instead; after all I can't expect that none of the classes that would use this have no other base class. Then I could use a separate `Local` and `Remote` trait, and some common base class to hide the details. I think I didn't want to have to declare all these type parameters twice.

I'll have another look at it.

Reply:
On second thought, I removed the `Local` and just pass `()` as the address in `InterpreterClient`. Less annoyance than having the duplicated code.

Reply:
And I suppose you can always create a field with an implementation of the class if inheritance is not an option, an inner class, so that's not a problem. Although the fact that the methods are `protected` could be inconvenient.